### ISSUE TITLE
Tagging browsers for simultaneous crawls 

### ISSUE THREAD

**Issue by hadiasghari** [openwpm/OpenWPM#114]
Consider the use case of crawling a long list of sites with browsers that have different configurations -- e.g. the 'do-not-track' parameter is set or unset. A simple script could be as follows:

    for site in sites:
            cs = CommandSequence.CommandSequence(site)
            cs.get()
            manager.execute_command_sequence(cs, index_label='DNT0')    
            manager.execute_command_sequence(cs, index_label='DNT1')  

Currently,  `execute_command_sequence()` can specify a command to be sent to the first available browser, one particular browser, or all browsers. There is however no way of specifying 'first available browser with a particular configuration', which would be ideal for this use case. A label set as an optional browser parameter could allow this.

**Comment by vringar** [comment]
@Akhila499 do you think this is something you would want to take on?

**Comment by Akhila499** [comment]
> @Akhila499 do you think this is something you would want to take on?

Thank you very much @vringar for offering this. I am happy to work on this. I did understand broadly as to what this feature is supposed to do. However, I do not know the exact details of this feature request. For instance where can I find the function execute_command_sequence()? Kindly direct me to the piece of code where this change needs to be made along with necessary documentation.

Thank you

**Comment by vringar** [comment]
Hey,
I'm would suggest you read through `demo.py` to get an overview of the crawler. You'll find [execute_command_sequence in the TaskManager](https://github.com/mozilla/OpenWPM/blob/master/automation/TaskManager.py#L565) but I don't think it'll be of much use to you without some context.
If you have any more general questions feel free to ask them on matrix (The link is at the top of the README.md)

**Comment by vringar** [comment]
My suggestion would be that the user can set an key called `labels` in the `browser_params` that is a list of all groups a certain browser belongs to.
`TaskManager.execute_command_sequence` would then also take a `label` as a parameter and then wait for the next browser with the right label to be ready and schedule the CommandSequence on there.

**Comment by vringar** [comment]
I'll get some cleanup for the code you'll have to edit in the next two days, so maybe wait for them to be landed before you start your coding and familiarize yourself with the core OpenWPM concepts until then.

**Comment by Akhila499** [comment]
@vringar I am on it. Will be going through demo.py and other documentation over the next couple of days before starting to code. Hope I can reach out to you in case I need any clarification.
Thanks

**Comment by vringar** [comment]
Hey, I saw that you asked for some more guidance in the channel.
Can you show me what you have done so far, so that I can be more helpful?

**Comment by Akhila499** [comment]
Hi @vringar Thanks for your reply. I have created new pull request (https://github.com/mozilla/OpenWPM/pull/585). I have not added anything. But, I guess I need to add a new key named 'label' and initialize it with empty list. However, I do not know how this affects the task manager and command sequence. Should I make any changes there as well? Let me know. Thanks

**Comment by vringar** [comment]
Hey,
then please go ahead and implement it as far as you understand it and see if you can maybe figure the rest out afterwards.
You don't need to change anything in the command sequence code and I have already linked you the method you'll need to change in the task manager.

**Comment by Akhila499** [comment]
Hi @vringar I was able to make a few changes. I added a key named "prioritize-this-config" and set it to False in default browser params. Also changed the code in TaskManager.execute_command_sequence. However, I do not see how the browser params flow into execute_command_sequence function. I tested it by passing an argument into the function execute_command_sequence directly in Demo.py. But, I am not sure if this is how it should work. I am not sure if this is indeed the actual feature request that was put forward. Please take a look at the changes I made. Thanks

Also, I am unable to debug properly with breaking points from demo.py (in Pycharm) as the instance browsers are going into some kind of waiting mode. Any inputs on this?

**Comment by vringar** [comment]
I execute_command_sequence you have a list of browsers each saving their respective browser_params
So you can do `self.browsers[i].browser_params` to access the config of the specific browser.

Can you tell me which part of my feature description you didn't understand?

> My suggestion would be that the user can set an key called `labels` in the `browser_params` that is a list of all groups a certain browser belongs to.
> `TaskManager.execute_command_sequence` would then also take a `label` as a parameter and then wait for the next browser with the right label to be ready and schedule the CommandSequence on there.

I don't quite understand your problem with pycharm as everything works as expected for me. Please consider filing another issue with the logs and a better description the problem

**Comment by Akhila499** [comment]
Hi @vringar Thanks for the input. I understood the way how to access a given browser's parmas. However, I am confused around the labels part that you mentioned. Can you elaborate on what you meant by the following sentence: 

"My suggestion would be that the user can set an key called labels in the browser_params that is a list of all groups a certain browser belongs to."

So, may I know this 'labels' key in each browser parameters, what does it exactly contain? You mentioned that the key 'labels' should contain a list of groups the given browser belongs to. However, I am not sure what these "groups" are. What about the values such as DNT0 and DNT1 @hadiasghari mentioned i.e. for the case of do-not-track? How your suggestion on groups and values such as DNT0 and DNT1 are related? Kindly elaborate.

Here is what I think: Please correct me if I am wrong. According to you, I just need to add all the keys that are enabled into the browser params into 'labels' list. For example, for browser 2, if both "adblock-plus" and "donottrack" are set to True then the labels category should look something like this: 
labels:['adblock-plus','donottrack']
Now, when the TaskManager.execute_command_sequence method is invoked with index_label argument as 'adblock-plus' which can look like as follows:

manager.execute_command_sequence(command_sequence, index_label='adblock-plus')

Then the program should wait until the browser that has adblock-plus config enabled is ready and then pick that browser (in this case browser 2) and start sending commands. This can be accomplished by checking 

self.browsers[i].browser_params['index_label]=='adblock-plus'

Also, I think the program should follow this path only if the other argument in execute_command_sequence 'index' is None.

This is what I understood. However, I am unable to reconcile how "DNT0" and "DNT1" kind of values fit in here. Is my assessment of the feature request wrong? Let me know. Thanks

**Comment by vringar** [comment]
`DNT0` and `DNT1` are labels just like `adblock-plus` and `donottrack`
I think you have a good understanding of this problem by now and agree with your statements. You should be able to produce some results based of that understanding.
If you still need further clarification I'd suggest that you consider dropping this issue and taking a look at another.

**Comment by hadiasghari** [comment]
Hi @vringar @Akhila499. I'd like to give an update on my request from 2016. I have since worked around this limitation—the inability to tell which Firefox instance or configuration relates to a particular visit—by running multiples crawl scripts each with only one (single) browser.  The crawl script instances (which I typically also run in  docker containers) configure the browser based on the script parameters, and save the sqlites in different directories, which means I can tell the visits apart. Of course, I then need to merge the sqlites (into one dataframe for analysis). But it gets the job done...

**Comment by Akhila499** [comment]
Hi @vringar @englehardt I made the changes. I also tested them. Let me know what you think.

**Comment by Akhila499** [comment]
@vringar Please let me know if I have to do something from my end

**Comment by vringar** [comment]
@Akhila499 Hey, I haven't heard back from you for a while and am now assuming that the PR is no longer under active development and this issue can be reassigned.

**Comment by Akhila499** [comment]
Hi @vringar I am still working on writing unit tests. will update you as soon as possible. Thanks for waiting.

**Comment by Akhila499** [comment]
Hi @vringar, Thank you very much for waiting. I have added testcases and fixed flake8 errors. Kindly check my PR https://github.com/mozilla/OpenWPM/pull/585 .Thanks a lot for guiding me through this process. I did learn a lot. Let me know.

Regards,
Akhila
